llvm/lib/IR/VFABIDemangler.cpp

```diff
@@ -566,7 +570,7 @@ FunctionType *VFABI::createFunctionType(const VFInfo &Info,
 
   auto *RetTy = ScalarFTy->getReturnType();
   if (!RetTy->isVoidTy())
-    RetTy = VectorType::get(RetTy, VF);
+    RetTy = ToWideTy(RetTy, VF);
```

Comment:
I think it's worth trying to pull out the VFABIDemangler changes (which also require `getContainedTypes`) into a separate PR. That would make it a bit easier to review the remaining vectoriser changes. I think logically they are independent changes because LoopVectorizationLegality will reject such functions anyway without this patch. There is already an existing unittests/Analysis/ReplaceWithVecLibTest.cpp unit test that you can hopefully adapt to cover the struct return cases.

Reply:
Split off the VFABI changes (and some helpers) into #119000 :+1: